dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/resolver/CategoryOptionResolver.java

```diff
@@ -0,0 +1,112 @@
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.resolver;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.hisp.dhis.expression.Expression.EXP_CLOSE;
+import static org.hisp.dhis.expression.Expression.EXP_OPEN;
+import static org.hisp.dhis.expression.Expression.SEPARATOR;
+import static org.hisp.dhis.expression.ParseType.INDICATOR_EXPRESSION;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.hisp.dhis.category.CategoryOption;
+import org.hisp.dhis.category.CategoryOptionStore;
+import org.hisp.dhis.common.DimensionalItemId;
+import org.hisp.dhis.expression.ExpressionService;
+import org.springframework.stereotype.Service;
+
+import com.google.common.base.Joiner;
+
+/**
+ * @author Dusan Bernat
+ */
+
+@Service( "org.hisp.dhis.analytics.resolver.CategoryOptionResolver" )
+public class CategoryOptionResolver implements ExpressionResolver
+{
+    private final ExpressionService expressionService;
+
+    private final CategoryOptionStore categoryOptionStore;
+
+    private static final String LEFT_BRACKET = "(";
+
+    private static final String RIGHT_BRACKET = ")";
+
+    private static final String CATEGORY_OPTION_PREFIX = "co:";
+
+    private static final String EMPTY_STRING = "";
+
+    public CategoryOptionResolver( ExpressionService expressionService, CategoryOptionStore categoryOptionStore )
+    {
+        checkNotNull( categoryOptionStore );
+        checkNotNull( expressionService );
+
+        this.expressionService = expressionService;
+        this.categoryOptionStore = categoryOptionStore;
+    }
+
+    @Override
+    public String resolve( String expression )
+    {
+        Set<DimensionalItemId> dimItemIds = expressionService.getExpressionDimensionalItemIds( expression,
+            INDICATOR_EXPRESSION );
+
+        Map<String, List<String>> resolvedOperands = new HashMap<>();
+
+        dimItemIds.forEach( id -> {
+            if ( id.getId1().contains( CATEGORY_OPTION_PREFIX ) )
+            {
+                CategoryOption co = categoryOptionStore
+                    .getByUid( id.getId1().replace( CATEGORY_OPTION_PREFIX, EMPTY_STRING ) );
+
+                String operand = EXP_OPEN + id.getId0() + SEPARATOR + id.getId1() +
+                    (id.getId2() != null ? SEPARATOR + id.getId2() : EMPTY_STRING) + EXP_CLOSE;
+
+                resolvedOperands.put( operand, co.getCategoryOptionCombos().stream()
+                    .map( c -> operand.replace( id.getId1(), c.getUid() ) ).collect( Collectors.toList() ) );
+            }
+        } );
+
+        for ( Map.Entry<String, List<String>> entry : resolvedOperands.entrySet() )
```

Comment:
I think a second loop here may be overkill, especially if `operand` is replaced with `id.getItem()`, but even if it wasn't. Instead of two loops it could just be one:

        for ( DimensionalItemId id : dimItemIds )
        {
            if ( id.getId1().contains( CATEGORY_OPTION_PREFIX ) )
            {
                CategoryOption co = categoryOptionStore
                    .getByUid( id.getId1().replace( CATEGORY_OPTION_PREFIX, EMPTY_STRING ) );

                if ( co != null )
                {
                    throw ...
                }

                List<String> resolved = co.getCategoryOptionCombos().stream()
                    .map( coc -> id.getItem().replace( id.getId1(), coc.getUid() ) )
                    .collect( Collectors.toList() );

                expression = expression.replace( id.getItem(),
                    LEFT_BRACKET + Joiner.on( "+" ).join( resolved ) + RIGHT_BRACKET );
            }
        }

To avoid a NPE you should throw an appropriate exception if the `co` is not found.

Also I suggest using `coc` as the variable inside the ap. I know that a single letter is usually used, but in this case, since we have concepts like Category and CategoryOption and CategoryOptionCombo, it helps reinforce to the reader that we are translating from `co` to `coc`.

Reply:
done